posix/fdtable.c

```diff
@@ -0,0 +1,245 @@
+#include <openenclave/enclave.h>
+#include <openenclave/internal/posix/device.h>
+#include <openenclave/internal/posix/fdtable.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/print.h>
+#include <openenclave/internal/thread.h>
+#include <openenclave/internal/trace.h>
+#include <openenclave/internal/utils.h>
+
+/*
+**==============================================================================
+**
+** Define the table of file-descriptor entries:
+**
+**==============================================================================
+*/
+
+/* The table allocation grows in multiples of the chunk size. */
+#define TABLE_CHUNK_SIZE 1024
+
+typedef oe_device_t* entry_t;
+
+static entry_t* _table;
+static size_t _table_size;
+static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
+static bool _installed_atexit_handler;
+
+static void _atexit_handler(void)
+{
+    /* Free the standard devices (but do not close them). */
+    for (size_t i = 0; i <= OE_STDERR_FILENO; i++)
+    {
+        oe_device_t* dev = _table[i];
+
+        if (dev)
+            oe_free(dev);
+    }
+
+    oe_free(_table);
+}
+
+static int _resize_table(size_t new_size)
+{
+    int ret = -1;
+
+    /* Install the atexit handler on the first call. */
+    if (!_installed_atexit_handler)
+    {
+        oe_atexit(_atexit_handler);
+        _installed_atexit_handler = true;
+    }
+
+    /* The fdtable cannot be bigger than the maximum int file descriptor. */
+    if (new_size > OE_INT_MAX)
+        goto done;
+
+    /* Round the new capacity up to the next multiple of the chunk size. */
+    new_size = oe_round_up_to_multiple(new_size, TABLE_CHUNK_SIZE);
+
+    if (new_size > OE_INT_MAX)
+        new_size = OE_INT_MAX;
+
+    if (new_size > _table_size)
+    {
+        entry_t* p;
+        size_t n = new_size;
+
+        /* Reallocate the table. */
+        if (!(p = oe_realloc(_table, n * sizeof(entry_t))))
+            goto done;
+
+        /* Zero-fill the unused portion. */
+        {
+            const size_t num_bytes = (n - _table_size) * sizeof(entry_t);
+
+            if (oe_memset_s(p + _table_size, num_bytes, 0, num_bytes) != OE_OK)
+                goto done;
+        }
+
+        _table = p;
+        _table_size = new_size;
+    }
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+/*
+**==============================================================================
+**
+** Public interface:
+**
+**==============================================================================
+*/
+
+int oe_fdtable_assign(oe_device_t* device)
+{
+    int ret = -1;
+    size_t index;
+
+    oe_spin_lock(&_lock);
+
+    if (!device)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* The file-descriptor table must be big enough for standard devices. */
+    if (_resize_table(TABLE_CHUNK_SIZE) != 0)
+        OE_RAISE_ERRNO(OE_ENOMEM);
+
+    /* Search for a free slot in the file descriptor table. */
+    for (index = OE_STDERR_FILENO + 1; index < _table_size; index++)
```

Comment:
`oe_dup` uses `oe_fdtable_assign` but this won't allow the [Redirecting Standard Output to a File](http://pubs.opengroup.org/onlinepubs/9699919799/functions/dup2.html) example in POSIX docs to work:

```
#include <unistd.h>
...
int pfd;
...
close(1);
dup(pfd);
close(pfd);
...
```

Reply:
Yeah, I noticed that too when I was writing the **dup2()** tests. I'm working on a new test for this.